plugins/samples/block_request/plugin.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_block_request]
+#include <string>
+
+#include "absl/strings/match.h"
+#include "proxy_wasm_intrinsics.h"
+
+// Checks the client referer and host headers match and
+// serves a 403 forbidden error if not.
+class MyHttpContext : public Context {
+ public:
+  explicit MyHttpContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers,
+                                       bool end_of_stream) override {
+    const auto referer = getRequestHeader("Referer");
+    const auto host = getRequestHeader("Host");
+    // Check if referer and host match.
+    if (referer && host && !absl::StrContains(referer->view(), host->view())) {
+      const auto requestId = generateRandomRequestId();
+      sendLocalResponse(403, "", "Forbidden - Request ID: " + requestId, {});
+      LOG_INFO("Forbidden - Request ID: " + requestId);
+      return FilterHeadersStatus::StopAllIterationAndWatermark;
+    }
+
+    addRequestHeader("Allowed", "true");
```

Comment:
Suggest choosing a header name such as "my-plugin-allowed" to make clear that this is not a standard HTTP header, and is added here just for illustrative purposes.

Reply:
Changed